controllers/hcpvaultsecretsapp_controller.go

```diff
@@ -357,8 +392,84 @@ func injectRequestInformation(runtime *httptransport.Runtime) {
 	runtime.Transport = &transport{child: runtime.Transport}
 }
 
+// getShadowSecretData retrieves shadowed secret data from a k8s secret
+func getShadowSecretData(ctx context.Context, k8sClient client.Client, o *secretsv1beta1.HCPVaultSecretsApp) (map[string]*models.Secrets20231128OpenSecret, error) {
+	// Get the shadow secret for the HCPVaultSecretsApp
+	shadowObjKey := makeShadowObjKey(o)
+	shadowSecret, err := helpers.GetSecret(ctx, k8sClient, shadowObjKey)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, nil
+		}
+		return nil, fmt.Errorf("failed to get shadow secret %s/%s: %w",
+			shadowObjKey.Namespace, shadowObjKey.Name, err)
+	}
+	// Decode shadowSecret.Data into a map of OpenSecret's
+	if shadowSecret != nil {
+		openSecrets := make(map[string]*models.Secrets20231128OpenSecret)
+		for k, v := range shadowSecret.Data {
+			secret, err := helpers.FromHVSShadowSecret(v)
+			if err != nil {
+				return nil, fmt.Errorf("failed to unmarshal shadow secret data %s/%s: %w",
+					shadowObjKey.Namespace, shadowObjKey.Name, err)
+			}
+			openSecrets[k] = secret
+		}
+		return openSecrets, nil
+	}
+
+	return nil, nil
+}
+
+func storeShadowSecretData(ctx context.Context, k8sClient client.Client, o *secretsv1beta1.HCPVaultSecretsApp, secrets []*models.Secrets20231128OpenSecret) error {
+	shadowObjKey := makeShadowObjKey(o)
+
+	// Delete the shadow secret if there are no secrets to store
+	if len(secrets) == 0 {
+		return helpers.DeleteSecret(ctx, k8sClient, shadowObjKey)
+	}
+
+	shadowSecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      shadowObjKey.Name,
+			Namespace: shadowObjKey.Namespace,
+			Labels:    shadowSecretLabels,
+		},
+	}
+	shadowSecret.ObjectMeta.Labels["hvs-app-name"] = o.Spec.AppName
+	shadowSecret.ObjectMeta.Labels["hcpvaultsecretsapp/name"] = o.Name
+	shadowSecret.ObjectMeta.Labels["hcpvaultsecretsapp/namespace"] = o.Namespace
+
+	shadowSecretData, err := helpers.MakeHVSShadowSecretData(secrets)
+	if err != nil {
+		return fmt.Errorf("failed to marshal shadow secret data %s/%s: %w",
+			shadowObjKey.Namespace, shadowObjKey.Name, err)
+	}
+	shadowSecret.Data = shadowSecretData
+
+	if err := helpers.CreateOrUpdateSecret(ctx, k8sClient, shadowSecret); err != nil {
```

Comment:
Is it possible for the shadowed secret to be orphaned when the HVSA instance is deleted?

Reply:
Right now the shadowed secret is deleted as part of [`handleDeletion()` in the controller](https://github.com/hashicorp/vault-secrets-operator/blob/3d2b4abbf93500dc28f1aca7769b006cf95a1b70/controllers/hcpvaultsecretsapp_controller.go#L360-L364), so if that fails then yes it could be orphaned. Did you have something else in mind?

Reply:
I see, could we add owner labels to the shadow secret? The owner could be the destination secret or the CR or both.

Reply:
We could use owner refs if we store the shadow secret in the CR's namespace; right now it's always in VSO's namespace.

I was a little hesitant to add something unexpected to the user's namespace, but it would certainly make cleanup more robust with owner refs.